File: algorithms/Searching/Medium/BreadthFirstSearch/VSDBreadthFirstSearch.py
# program to traverse the graph using BFS
# Function takes graph as a input parameters

def BFS(graph):
    queue=[] # Is used to implement the BFS logic
    visited=[] # Is used to store visited vertices
    first = list(graph.keys())[0]
    queue.append(first)
    while queue:
        for vertex in graph[queue[0]]:
            if vertex not in visited and vertex not in queue:
                queue.append(vertex)
        visited.append(queue.pop(0))
    return visited
    

# Test drive code:
graph={
    0:[1,2],
    1:[2],
    2:[0,3],
    3:[3]

}
print("The BFS traversal of the graph is :",BFS(graph))

File: algorithms/Searching/Medium/BreadthFirstSearch/VSDBreadthFirstSearch.cpp
#include<bits/stdc++.h>

using namespace std;


class Graph{
    
    public:
        int size;
        vector<vector<int>> graph;
        Graph(int v){
            size=v;
            graph.resize(v);
        }
        void addEdge(int v,int child){
            graph[v].push_back(child);
        }

        void bfs(int start){
            queue<int> q;
            vector<bool> visited(size,false);
            q.push(start);
            visited[start]=true;
            while(!q.empty()){    
                        
               start = q.front();
               q.pop();
               cout<<start<<" ";
                for(auto it:graph[start]){
                    
                    if(!visited[it]){
                        visited[it]=true;
                        q.push(it);
                    }
                }

            }

        }

};

int main(){
    Graph g(4);
    g.addEdge(0,1);
    g.addEdge(0,2);
    g.addEdge(1,2);
    g.addEdge(2,0);
    g.addEdge(2,3);
    g.addEdge(3,3);
    g.bfs(2);
    return 0;
}